Add "indeterminate" prop to <Checkbox>

nucleartide opened the pull request:
Closes #312.

<img width="1117" alt="screen shot 2018-08-30 at 4 52 00 pm" src="https://user-images.githubusercontent.com/914228/44885330-12b80100-ac75-11e8-83f9-1857d5f163b0.png">

vadimdemedes commented on src/checkbox/stories/index.stories.js:
```diff
@@ -5,9 +5,42 @@ import { Checkbox } from '../../checkbox'
-    <Checkbox checked label="Checkbox checked" />
-    <Checkbox disabled label="Checkbox disabled" />
-    <Checkbox disabled checked label="Checkbox checked disabled" />
+    {permutations({
```
What's the benefit of having this? I would just add:

```jsx
<Checkbox indeterminate label="Checkbox indeterminate" />
```

vadimdemedes commented on src/checkbox/stories/index.stories.js:
```diff
@@ -9,5 +9,6 @@ storiesOf('checkbox', module).add('Checkbox', () => (
     <Checkbox checked label="Checkbox checked" />
     <Checkbox disabled label="Checkbox disabled" />
     <Checkbox disabled checked label="Checkbox checked disabled" />
+    <Checkbox checked indeterminate label="Checkbox checked disabled" />
```
Label should be `checked indeterminate`

nucleartide replied:
Fixed!

nucleartide commented:
@jeroenransijn Updated!

jeroenransijn commented on src/checkbox/src/Checkbox.js:
```diff
@@ -119,10 +147,11 @@ class Checkbox extends PureComponent {
           onChange={onChange}
           disabled={disabled}
           {...(isInvalid ? { 'aria-invalid': true } : {})}
+          ref={this.setIndeterminate}
```
`ref` is actually not available on `Box` since it's a React component — instead `Box` uses the non-standard but conventional `innerRef` prop. 

Also, this might actually not work completely since I think `ref/innerRef` is only called once in the lifecycle of the component. We might need to opt-in to `componentDidUpdate`: https://reactjs.org/docs/react-component.html#componentdidupdate

nucleartide commented:
@jeroenransijn Addressed your comment about `ref` + `componentDidUpdate`. With `innerRef`, it seems to work when you pass in new props.

Try this component in the story:

```jsx
import React from 'react'

class CheckboxWithState extends React.Component {
  state = {
    checked: false,
  }

  handleChange = () => {
    this.setState(() => ({ checked: !this.state.checked }))
  }

  render() {
    return <Checkbox checked={this.state.checked} onChange={this.handleChange} />
  }
}
```
